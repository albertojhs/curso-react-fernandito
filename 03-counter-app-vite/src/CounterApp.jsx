import { useState } from 'react';
import PropTypes from "prop-types";


export const CounterApp = ({ value }) => {
    
    /* Declaración de constantes */
    const [ counter, setCounter ] = useState( value );


    

    /*Declaración de funciones */
    const handleAdd = () => { 
        setCounter( counter + 1 ); 
    }

    const handleSub = () => {
        setCounter( counter - 1 );
    }

    const handleReset = () => {
        setCounter( value );
    }




    /*Return para el render */
    return (
        <>
            <h2>CounterApp</h2>
            <h3> { counter } </h3>
            <button onClick={ handleAdd }>
                +1
            </button>
            <button onClick={ handleSub }>
                -1
            </button>
            <button aria-label="btn-reset" onClick={ handleReset }>
                reset
            </button>
        </>
    );
} 



/*Tipado, restricciones, etc del props */
CounterApp.propTypes = {
    value : PropTypes.number.isRequired
}