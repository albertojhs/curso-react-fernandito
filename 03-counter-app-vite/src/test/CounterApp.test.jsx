import { fireEvent, render, screen } from '@testing-library/react';
import { CounterApp } from '../CounterApp';

describe('Pruebas para CounterApp', () => { 

    test('Debe de coincidir con el snapshoot', () => { 
        const { container } = render( <CounterApp value={ 100 }/> );
        expect( container ).toMatchSnapshot();
    });

    test('Debe de mostrar el valor inicial de 100', () => {
        render( <CounterApp value={100} /> );
        expect( screen.getByRole('heading', {level: 3}).innerHTML.trim() ).toEqual( '100' );
        
        //Versión de Fernandito
        //Con esta versión hay que tener mucho cuidado ya que debemos estar seguros que solo existe un 
        //100 en toda la página
        expect( screen.getByText( '100' ) ).toBeTruthy();
    });


    test('Debe de incrementar con el evento +1', () => {
        render( <CounterApp value={ 10 } /> );
        fireEvent.click( screen.getByText('+1') );
        expect( screen.getByText('11') ).toBeTruthy();
    });


    test('Debe decrementar con el evento -1', () => {
        render( <CounterApp value={ 10 } /> );
        fireEvent.click( screen.getByText( '-1' ) );
        expect( screen.getByText( '9' ) ).toBeTruthy();
    });
    
    
    test('Debe regresar al valor inicial con el evento reset', () => {
        render( <CounterApp value={ 10 } /> );
        fireEvent.click( screen.getByText( '-1' ) );
        fireEvent.click( screen.getByText( 'reset' ) );
        expect( screen.getByText( '10' ) ).toBeTruthy();
        
        //Manera 2, usando aria-label en el DOM para llamarlo por el name, OJO usar la propiedad
        //name directamente no funciona, debe ser aria-label
        fireEvent.click( screen.getByText( '+1' ) );
        fireEvent.click( screen.getByRole( 'button', {name: 'btn-reset'} ) );
        expect( screen.getByText( '10' ) ).toBeTruthy();
        
        
    });

 })