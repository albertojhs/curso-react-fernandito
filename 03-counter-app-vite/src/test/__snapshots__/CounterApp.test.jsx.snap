// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas para CounterApp Debe de coincidir con el snapshoot 1`] = `
<div>
  <h2>
    CounterApp
  </h2>
  <h3>
     
    100
     
  </h3>
  <button>
    +1
  </button>
  <button>
    -1
  </button>
  <button
    aria-label="btn-reset"
  >
    reset
  </button>
</div>
`;
